### PR TITLE
Enable analyzing nested input- and output-dicts

### PR DESCRIPTION
**Pull request by snimu**
Fixes [issue#141](https://github.com/TylerYep/torchinfo/issues/141).

There was an unexpected complication, as described in [my comment on pull-request#195](https://github.com/TylerYep/torchinfo/pull/195#issuecomment-1381463098): some `num_params` were negative. I've fixed it by ignoring negative numbers for `total_params` and `trainable_params` in `ModelStatistics.__init__(...)`. 

I don't know why the `num_params` is sometimes negative, and so cannot say for sure that this solution is correct, only that it produces consistent results. Hopefully, this doesn't cause any issues in the future, though at some point, I would like to experimentally validate at least the calculated memory consumption.

Speaking of memory consumption, neither `LayerInfo.param_bytes` nor `LayerInfo.output_bytes` is ever negative, at least not in the models that I've tested, so I haven't included corresponding checks for this pull-request.

@mert-kurttutan stated in [his comment to pull-request#195](https://github.com/TylerYep/torchinfo/pull/195#issuecomment-1310645452) that it might be interesting to accumulate the sizes of all `torch.Tensor` in a nested structure. I have chosen not to implement this, because it leads to a few complications:

1. If the resulting list of sizes is nested (e.g. `[[3, 64, 64], [3, 64, 64]]`), `torchinfo.summary` fails (I've tried this; IIRC, `raise ValueError("Unknown input type")` is triggered in `torchinfo.py::forward` because of `x`, though I don't know why). Some rewriting would have to occur. 
2. If the resulting list of sizes is flattened (e.g. `[3, 64, 64, 3, 64, 64]`), the output of `summary` becomes very ugly very quickly. Also, this would be confusing, as it is now unclear what the exact input- and output-structure is.

In both cases, backward-compatability would be broken in some cases (the output of `summary` would look different for some models & inputs), so a decision would have to be made whether this is acceptable, or a new parameter (`recurse_inputs` or something like that) would have to be added to give the user control over this feature.

I didn't want to make the design-decisions and thought that this basic solution for making `dict` work as well for `input_data` as it already works for `tuple` and `list` is sufficient for the time being. 

Thoughts appreciated :)

### CODE REVIEW

File: torchinfo/model_statistics.py [hunk 1]
```diff
@@ -33,14 +33,24 @@ def __init__(
...
+                    else 0
+                )
             else:
                 if layer_info.is_recursive:
                     continue
-                self.total_params += layer_info.leftover_params()
-                self.trainable_params += layer_info.leftover_trainable_params()
+                leftover_params = layer_info.leftover_params()
+                leftover_trainable_params = layer_info.leftover_trainable_params()
+                self.total_params += leftover_params if leftover_params > 0 else 0
```
Comment (TylerYep) [inline torchinfo/model_statistics.py:50]: This is suspicious, let's see if we can figure out why params can ever be negative. I'll take a look too

**Review by TylerYep — COMMENTED** [review]
Overall change seems good. My only feedback is that we should try to get some test coverage for the missing lines to make sure they work properly.

**Comment by snimu** [comment]
Alright, I'll increase coverage 🙂

**Comment by snimu** [comment]
Increased coverage (though I'm working on more - don't merge yet); 

replaced the custom torch version comparison; I don't quite understand what you mean by this:

> This also makes my test.yml file a lot simpler, thanks for the suggestion

**Comment by snimu** [comment]
When writing tests for this PR, I have had trouble creating inputs to `LayerInfo.calculate_size` that would cover an edgecase but not raise an exception somewhere else in torchinfo, so I created [torch-nested](https://github.com/snimu/torch-nested) to externalize the functionality and tests. It seems to work very well and would have the following advantages:

1. Make test-coverage easier for torchinfo (torch-nested currently has [code-coverage of 99.6%](https://app.codecov.io/gh/snimu/torch-nested))
2. Make `LayerInfo.calculate_size` more readable and maintainable
3. For the future, make it easier to extend torchinfo to accept more data-structures. 

I'm actively working on torch-nested, but the basic functionality needed for torchinfo&mdash;easy accessing of tensors in deeply nested structures&mdash;will not change, but only be extended to more data-structures.

So far, I've replicated the functionality of `LayerInfo.calculate_size` one-to-one. I don't quite understand why the last 
`torch.Tensor` is used for `dict` but the first for `list` and `tuple`, but have preserved that functionality, too.

What do you think?

**Comment by snimu** [comment]
Btw, the additional cov-miss is coming from the `except ImportError`-statement in line 14 of layer_info.py. Unfortunately, this is necessary for the reason named in the comment:

```
# Repeated imports: if "import torch" etc. are outside try-except,
    #   isort will put them above the "from torch_nested ..."-import,
    #   which will cause pylint to throw the following error:
    # "C0412: Imports from package torch are not grouped (ungrouped-imports)"
```

**Comment by TylerYep** [comment]
Hi, unfortunately I am not willing to accept PRs moving parts of torchinfo to an outside library and using that library instead - as a maintainer it will add a lot of additional burden in terms of aligning releases, version support, and ease of refactoring without breaking existing users' code unintentionally.

You can bring the necessary parts of torch-nested library's functionality to torchinfo and I can review (which I believe was your previous version of this PR), but adding that library as a dependency is a nonstarter.

> trouble creating inputs to LayerInfo.calculate_size that would cover an edgecase but not raise an exception somewhere else in torchinfo

If this is the case then we should address this exceptions, please feel free to share the code that created these exceptions and we can investigate and fix those issues.

> I don't quite understand why the last torch.Tensor is used for dict but the first for list and tuple, but have preserved that functionality, too.

This was arbitrary and I am willing to make a backwards incompatible change to this, as long as it makes the behavior better. Ideally we should show all tensor sizes in dicts but if we can only show one for now then any tensor is fine - this behavior is not set in stone yet.

**Comment by snimu** [comment]
> Hi, unfortunately I am not willing to accept PRs moving parts of torchinfo to an outside library and using that library instead - as a maintainer it will add a lot of additional burden in terms of aligning releases, version support, and ease of refactoring without breaking existing users' code unintentionally.

Alright, I understand. 

> Ideally we should show all tensor sizes in dicts ...

I've played with extracting the full shape of nested structures in `LayerInfo.calculate_size`, but other parts of torchinfo had problems with the output&mdash;specifically, somewhere in `forward_pass`, an exception occured with very nested structures (it may have been `process_input` instead, it was a week ago ;)This is actually why I thought creating the torch-nested package would make sense, to create a generalized way of interacting with nested data-structures containing `torch.Tensors`. However, I fully understand that you don't want the burden of having to interact with external packages, so I've fixed the issue without using it now. 

> ... but if we can only show one for now then any tensor is fine - this behavior is not set in stone yet.

I have left it in for now. Thought it might be better to change the behavior in one go when the change is made to show the full data-structure-size. 

> > trouble creating inputs to LayerInfo.calculate_size that would cover an edgecase but not raise an exception somewhere else in torchinfo

> If this is the case then we should address this exceptions, please feel free to share the code that created these exceptions and we can investigate and fix those issues.

I think I may have mixed up some things: IIRC, I've mostly raised exceptions when trying to extract the full size and shape of nested data-structures as described above, not for test-cases. Sorry. If I have time, I'll look into it at some point.

**Comment by snimu** [comment]
Hi @TylerYep, this comment is just to make sure that you didn't miss my last push & comment for this PR under all the other comments etc. :) It:

- Solved [issue#141](https://github.com/TylerYep/torchinfo/issues/141) and (I think) [issue#215](https://github.com/TylerYep/torchinfo/issues/215)
- Increased test coverage of torchinfo
- Preserved all other behavior

If you've just been to busy, ignore this.

**Comment by snimu** [comment]
Now fixes [issue#215](https://github.com/TylerYep/torchinfo/issues/215) in addition to [issue#141](https://github.com/TylerYep/torchinfo/issues/141) (and with good test-coverage)

**Comment by TylerYep** [comment]
Looks great, thank you for fixing this and cleaning up the code along the way! This code can be really confusing and it's great that this solution can work for many models and external libraries too.

**Pull request merged**